skeletons/bner_support.c

```diff
@@ -0,0 +1,369 @@
+                    (uint8_t)(tag_lvt.u.length >> 8) & 0xff;
+                buf[required_size + 3] = (uint8_t)tag_lvt.u.length & 0xff;
+            }
+            required_size += 4;
+        }
+        break;
+    case BNER_LVT_VALUE:
+        if(size) buf[0] |= tag_lvt.u.value;
+        break;
+    case BNER_LVT_TYPE:
+        if(size) buf[0] |= tag_lvt.u.type;
+        break;
+    }
+
+    return required_size;
+}
+
+ber_tlv_tag_t
+convert_ber_to_bner_tag(ber_tlv_tag_t ber_tag) {
+    /*
+     * The asn1c compiler will produce UNIVERSAL tags for primative types
+     * that need to be translated into APPLICATION tags for BACnet encoding
+     * rules
+     */
+
+    /* clang-format off */
+    /* -- ***************** Application Types ********************
+     * -- The following productions are the definitions of the Application datatypes.
+     * -- See Clause 20.2.1.4.
+     * -- NULL			[APPLICATION 0], equivalent to [UNIVERSAL 5]
+     * -- BOOLEAN		[APPLICATION 1], equivalent to [UNIVERSAL 1]
+     * Unsigned ::=		[APPLICATION 2] INTEGER (0..MAX)
+     * Unsigned8 ::=		Unsigned (0..255)
+     * Unsigned16 ::=		Unsigned (0..65535)
+     * Unsigned32 ::=		Unsigned (0..4294967295)
+     * -- INTEGER		[APPLICATION 3], equivalent to [UNIVERSAL 2]
+     * INTEGER16 ::=		INTEGER (-32768..32767)
+     * -- REAL			[APPLICATION 4], equivalent to [UNIVERSAL 9] ANSI/IEEE-754 single precision floating point
+     * Double ::=		[APPLICATION 5] OCTET STRING (SIZE(8)) -- ANSI/IEEE-754 double precision floating point
+     * -- OCTET STRING		[APPLICATION 6], equivalent to [UNIVERSAL 4]
+     * CharacterString ::=	[APPLICATION 7] OCTET STRING -- see Clause 20.2.9 for supported types
+     * -- BIT STRING		[APPLICATION 8], equivalent to [UNIVERSAL 3]
+     * -- ENUMERATED		[APPLICATION 9], equivalent to [UNIVERSAL 10]
+     * Date ::=		[APPLICATION 10] OCTET STRING (SIZE(4)) -- see Clause 20.2.12
+     * --	first octet	year minus 1900,		X'FF' = unspecified
+     * --	second octet	month		(1.. 14),	1 = January
+     * --							13 = odd months
+     * --							14 = even months
+     * --							X'FF' = unspecified
+     * --	third octet	day of month	(1..34),	32 = last day of month
+     * --							33 = odd days of month
+     * --							34 = even days of month
+     * --							X'FF' = unspecified
+     * --	fourth octet	day of week	(1..7)		1 = Monday
+     * --							7 = Sunday
+     * --							X'FF' = unspecified
+     * Time ::=		[APPLICATION 11] OCTET STRING (SIZE(4)) -- see Clause 20.2.13
+     * --	first octet	hour		(0..23),	X'FF' = unspecified
+     * --	second octet	minute		(0..59),	X'FF' = unspecified
+     * --	third octet	second		(0..59),	X'FF' = unspecified
+     * --	fourth octet	hundredths	(0..99)		X'FF' = unspecified
+     * BACnetObjectIdentifier ::= [APPLICATION 12] OCTET STRING (SIZE(4)) -- see Clause 20.2.14
+     */
+    /* clang-format on */
+
+    ber_tlv_tag_t bner_tag;
+
+    if(ber_tag == -1) return -1; /* ANY */
```

Comment:
@vlm 
I have this check here because an asn1 rule like this:
```asn1
BACnetTimeValue ::= SEQUENCE {
        time            Time,
        value           ABSTRACT-SYNTAX.&Type  -- any primitive datatype; complex types cannot be decoded
        }
```
will produce code like this:
```c
asn_TYPE_member_t asn_MBR_BACnetTimeValue_1[] = {
	{ ATF_NOFLAGS, 0, offsetof(struct BACnetTimeValue, time),
		(ASN_TAG_CLASS_APPLICATION | (11 << 2)),
		0,
		&asn_DEF_Time,
		0,
		{ 0, 0, 0 },
		0, 0, /* No default value */
		"time"
		},
	{ ATF_ANY_TYPE | ATF_NOFLAGS, 0, offsetof(struct BACnetTimeValue, value),
		-1 /* Ambiguous tag (ANY?) */,
		0,
		&asn_DEF_ANY,
		0,
		{ 0, 0, 0 },
		0, 0, /* No default value */
		"value"
		},
```
The line in the code above:
```c
		-1 /* Ambiguous tag (ANY?) */,
```
is of type `ber_tlv_tag_t`, which is `typedef unsigned ber_tlv_tag_t`

However, this line of code causes this compilation error:
```
bner_support.c: In function ‘convert_ber_to_bner_tag’:
bner_support.c:320:16: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     if(ber_tag == -1) return -1; /* ANY */
                ^
```

I tried changing it to `typedef signed ber_tlv_tag_t`, but this causes `tests/tests-skeletons/check-ber_tlv_tag` to fail:
```
$ cat tests/tests-skeletons/test-suite.log
========================================================
   asn1c 0.9.29: tests/tests-skeletons/test-suite.log
========================================================

# TOTAL: 17
# PASS:  16
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: check-ber_tlv_tag
=======================

../../../tests/tests-skeletons/check-ber_tlv_tag.c:103:21: runtime error: left shift of negative value -294967296
    #0 0x401003 in main (/home/jringle-admin/git/asn1c/.tmp.build/tests/tests-skeletons/check-ber_tlv_tag+0x401003)
    #1 0x2aef36a22f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #2 0x40115f  (/home/jringle-admin/git/asn1c/.tmp.build/tests/tests-skeletons/check-ber_tlv_tag+0x40115f)
```

Do you have any suggestions?

Reply:
It seems to me that X.690 section 8.1.2 doesn't seem to specify a maximum tag value allowed, but rather that the tag value can be arbitrarily large just by adding more subsequent identifier octets with the high order bit set. As such, it seems to me that the limit on the largest tag value would be limited by the implementation that is implementing the Basic Encoding Rules. Are there users of BER out there where you could reasonably expect that they would experience hardship if the largest BER tag value was cut in half from this implementation's max value due to changing it to `typedef signed ber_tlv_tag_t`?

Reply:
I have not seen it, but it changing signedness of ber_tlv_tag_t will surely cause me going over the code base for quite some time in search of inadvertent security issues. So do it as a separate PR to make it a bit easier.

Reply:
@vlm I abandoned the idea of changing it to signed and instead did https://github.com/vlm/asn1c/commit/570027752233bb3016bdaf1658c1f2b5043397fc